src/pcre2_compile.c

```diff
@@ -11147,6 +11147,8 @@ an offset is available in the parsed pattern. */
 ptr = pattern + cb.erroroffset;
 
 HAD_EARLY_ERROR:
+PCRE2_ASSERT(ptr >= pattern); /* Ensure we don't return invalid erroroffset */
+PCRE2_ASSERT(ptr < (pattern + patlen));
```

Comment:
`s/</<=/`, since the NUL at the end of pattern is a perfectly valid erroroffset

Reply:
@carenas Thanks for mentioning that. I actually flip-flopped between `<` and `<=` when preparing this PR. However, I noticed that (**if** I have understood well) it seems that patterns which are passed to `pcre2_compile` are not always NUL-terminated. (When the pattern is NUL-terminated, one can pass `PCRE2_ZERO_TERMINATED` as the `length` argument to `pcre2_compile`.)

Is that correct? If so, should the assertion be predicated on `length == PCRE2_ZERO_TERMINATED`?

Reply:
> should the assertion be predicated on `length == PCRE2_ZERO_TERMINATED`?

no; and it is unlikely to work as the length value is reset earlier AFAIK.

the [documentation](https://pcre2project.github.io/pcre2/doc/html/pcre2api.html) mentions:

> Some errors are not detected until the whole pattern has been scanned; in these cases, the offset passed back is the length of the pattern

Reply:
> > should the assertion be predicated on `length == PCRE2_ZERO_TERMINATED`?
> 
> no; and it is unlikely to work as the length value is reset earlier AFAIK.

I think I didn't express myself well. I was referring to the value of `length` *on function entry*, not at the point where the assertion occurs.

> the [documentation](https://pcre2project.github.io/pcre2/doc/html/pcre2api.html) mentions:
> 
> > Some errors are not detected until the whole pattern has been scanned; in these cases, the offset passed back is the length of the pattern

Hmm. If that is so, then I guess `pattern + erroroffset` is not guaranteed to be a valid pointer. Since PCRE2 users may presumably wish to use `pattern + erroroffset` to display the part of the pattern which caused an error, this feels like a defect in the API to me.

I guess if users read the documentation, then they should know to check for the special value `erroroffset == patlength`.

Reply:
Anyways, regardless of the issues mentioned, it is apparent that the assertion should indeed use `<=` and not `<`. PR updated.

Reply:
> I guess if users read the documentation, then they should know to check for the special value `erroroffset == patlength`.

sure hope they do, specially if trying to print the place where the pattern error was found, as they then would realize it might not be printable, if it happens to be in the middle of an UTF8 character, for example.

Reply:
> > > should the assertion be predicated on `length == PCRE2_ZERO_TERMINATED`?
> > 
> > 
> > no; and it is unlikely to work as the length value is reset earlier AFAIK.
> 
> I think I didn't express myself well. I was referring to the value of `length` _on function entry_, not at the point where the assertion occurs.

What I meant was that there is no way for PCRE2 to know if a pattern is NUL terminated or not (PCRE2_ZERO_TERMINATED is only a hint and an "optimization" so callers of this function can rely on PCRE2 doing `strlen` for them).

For additional context, '\0' in patterns used to be impossible before PCRE2.